### PR TITLE
MM 19654 s3 debugging logs overlap, Wrapped mlog in an io.Writer func

### PR DESCRIPTION
**Pull request by TealWater**
#### Summary

There was an issue of duplicate logs being produced when s3 enabled their debug logs.

The goal of this fix is to combine the s3 logs into one global logger, therefore preventing duplicate messages.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost/issues/19654
#### Release Note

```release-note
none
```

### CODE REVIEW

File: server/platform/shared/filestore/s3store.go [hunk 1]
```diff
@@ -178,7 +186,8 @@ func (b *S3FileBackend) s3New(isCloud bool) (*s3.Client, error) {
 	}
 
 	if b.trace {
-		s3Clnt.TraceOn(os.Stdout)
+		s3Trace := s3Trace{}
```
Comment (agnivade) [inline server/platform/shared/filestore/s3store.go:189]: It's weird to have variable name and struct name as the same. Let's name the variable as `tracer` or sth similar.

File: server/platform/shared/filestore/s3store.go [hunk 2]
```diff
@@ -124,6 +124,14 @@ func newS3FileBackend(settings FileBackendSettings, isCloud bool) (*S3FileBacken
 	return backend, nil
 }
 
+type s3Trace struct {
+}
+
+func (sTrace *s3Trace) Write(in []byte) (int, error) {
```
[the remark this answers is not in the document]
Comment (TealWater) [reply server/platform/shared/filestore/s3store.go:130]: I'm sorry.... receiver name? Do you mean:
```
var tracer s3Trace = new &s3Trace{}
num , _ := tracer.Write(bytes)
```

File: server/platform/shared/filestore/s3store.go (hunk 1 above)
Comment (TealWater) [reply server/platform/shared/filestore/s3store.go:189]: This seems fair, my bad.

File: server/platform/shared/filestore/s3store.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (TealWater) [reply server/platform/shared/filestore/s3store.go]: I implemented the latter.

[the remark this answers is not in the document]
Comment (agnivade) [reply server/platform/shared/filestore/s3store.go]: Yeah the latter is called the receiver. https://go.dev/ref/spec#Receiver

**Review by agnivade — APPROVED** [review]
Looks good!

Looks like golangci is complaining to remove the `_` altogether:

```
Error: platform/shared/filestore/s3store.go:130:1: receiver-naming: receiver name should not be an underscore, omit the name if it is unused (revive)
func (_ *s3Trace) Write(in []byte) (int, error) {
	mlog.Debug(string(in))
	return len(in), nil
}
```

File: server/platform/shared/filestore/s3store.go [hunk 3]
```diff
@@ -178,7 +186,8 @@ func (b *S3FileBackend) s3New(isCloud bool) (*s3.Client, error) {
 	}
 
 	if b.trace {
-		s3Clnt.TraceOn(os.Stdout)
+		s3Tracer := s3Trace{}
+		s3Clnt.TraceOn(&s3Tracer)
```
Comment (hanzei) [inline server/platform/shared/filestore/s3store.go:190]: Nit pick:
```suggestion
		s3Clnt.TraceOn(&s3Tracer{})
```

**Review by hanzei — APPROVED** [review]
Great work :+1: 

Thanks for the contribution @TealWater

File: server/platform/shared/filestore/s3store.go (hunk 3 above)
Comment (TealWater) [reply server/platform/shared/filestore/s3store.go:190]: LMAO, yes captain.

**Review by hanzei — APPROVED** [review]

**Comment by agnivade** [comment]
@DHaussermann - The test steps are to enable `FileSettings.AmazonS3Trace` and confirm that the trace logs flow through the MM logger rather than just dump to stdout.

**Review by DHaussermann — APPROVED** [review]
Tested and passed. 
Enabled `AmazonS3Trace` and setup Amazon S3 bucket deliberately hitting various misconfigurations. 

All logs looked as expected in the normal format. I saw no duplication. 

LGTM!

Please merge.

**Pull request merged**